### ISSUE TITLE
kde applications fail to open file dialog 

### ISSUE THREAD

**Issue by kensully** [flatpak/xdg-desktop-portal#168]
Hi, I am on Ubuntu 16.04.4 platform. I installed `org.kde.kwordquiz` from `kdeapps` repo according to guide in this link: https://community.kde.org/Guidelines_and_HOWTOs/Flatpak

It launches well but it doesn't open file browser to select new word list. 
I get the following error:
         OpenURI portal is not available
Would you please help me fix it?

**Comment by mariospr** [comment]
You need to have the xdg-desktop-portal package installed to have access to the OpenURI portal, as well as a backend such as xdg-dekstop-portal-gtk to be actually able to use it (although in your case I'm guessing you'll probably want [xdg-desktop-portal-kde](https://phabricator.kde.org/source/xdg-desktop-portal-kde/), although I'm unaware of what the current status of it is)

**Comment by kensully** [comment]
@mariospr thank you for responding. I have installed all the packages you mentioned including `xdg-desktop-portal-kde`. But I have a specific condition, I enabled `multiarch` support in my Ubuntu distro, then I switched to `amd64` kernel in i386 based installation. I don't know if that matters.

**Comment by mariospr** [comment]
I don't see how that would matter to this, seems like unrelated to me. I think what I would do now is try to debug this a little bit more...

What happens if you use `dbus-monitor` or `bustle`? Can you see the messages from the sandboxed application to the OpenURI portal flying by? Is there anyone listening / owning the bus name for the OpenURI portal? Are there the processes for xdg-desktop-portal and xdg-desktop-portal-[kde|gtk] running?...

Also, check the journal as well. It could be that you had the processes for the portal running but that for some reason (e.g. a crash) they exit before they could handle the request.

Last, see below a command that might be useful to help you debug this, if you run it from **inside the sandbox**, at least with a GNOME runtime:
```
gjs -c "imports.gi.Gio.AppInfo.launch_default_for_uri('http://www.flatpak.org', null)"
```

(To get a shell into the sandbox, you can do something like `flatpak run org.gnome.Platform/x86_64/3.26`, for that runtime. I'm not sure if that will work with a KDE runtime, though, but for debugging purposes I think it's still interesting that you try with a GNOME runtime first)

**Comment by kensully** [comment]
You are right, I tried both ` flatpak run org.gnome.Platform/x86_64/3.26` and ` flatpak run org.gnome.Platform/i386/3.26`, 

I get :    

       sh-4.3$ gjs -c "imports.gi.Gio.AppInfo.launch_default_for_uri('http://www.flatpak.org', null)"

      (gjs:5): GLib-GIO-WARNING **: Cannot create document portal proxy: Error calling StartServiceByName for org.freedesktop.portal.Desktop: GDBus.Error:org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.portal.Desktop exited with status 127

      (gjs:5): Gjs-WARNING **: JS ERROR: Gio.IOErrorEnum: OpenURI portal is not available
      @<command line>:1:1

      JS_EvaluateScript() failed

**Comment by kensully** [comment]
I found the path of `xdg-desktop-portal-kde`, it's `/usr/local/lib/i386-linux-gnu/libexec/xdg-desktop-portal-kde`

**Comment by kensully** [comment]
One more point to add, I installed all `flatpak` and friends from source, and  all the pakages is installed in /usr/local`

**Comment by mariospr** [comment]
Hmm.. which version of the portals do you have? I'm not 100% sure if this might be related, but a recent change in flatpak moved the document portal from flatpak itself to xdg-desktop-portal, and you error just rang that bell in my head.

As mentioned in the release email for flatpak 0.10.4:
```
This version adds the --disable-document-portal configure option, which is needed when
packaging xdg-desktop-portal 0.10 which now ships the document portal instead. Anyone
packaging this new version need to disable the portal in flatpak and add a dependency
on xdg-desktop-portal from the flatpak package.
```

So, if you're building everything from sources as you said, I understand you're building xdg-desktop-portal >= 0.10 and flatpak >= 0.10.4, in which case you'd need to pass `--disable-document-portal` when building flatpak or bad things will happen (maybe *these* ones?)

PS: About installing everything in `/usr/local`, that should not be a problem as long as you don't have flatpak + portals also installed from your distro, under `/usr` (which may or may not cause trouble too). Just to be sure, I think I'd check `ps` to see if there's any binary running that is not the one you expect.

**Comment by hadess** [comment]
> PS: About installing everything in /usr/local, that should not be a problem as long as you don't have flatpak + portals also installed from your distro, under /usr (which may or may not cause trouble too). Just to be sure, I think I'd check ps to see if there's any binary running that is not the one you expect.

That's probably the problem actually, with D-Bus/systemd unable to find the services files under `/usr/local`. Ideally, you'd want somebody to backport those packages to your current version.

**Comment by kensully** [comment]
@mariospr I have `Flatpak 0.11.1`. Yes, I installed all the packages from `github`. I 'll try to build `flatpak` with that configuration and let you know.
@hadess  thank you for suggestion, I'll also keep that in mind.

**Comment by mariospr** [comment]
> That's probably the problem actually, with D-Bus/systemd unable to find the services files under /usr/local. Ideally, you'd want somebody to backport those packages to your current version.

Me an my uncanny ability to misjudge situations... @hadess Now you mention I think you're right, I was not thinking of the problem finding D-Bus services if installed there, but it makes all sense.

@kensully I'd still suggest you check whether you need to compile disabling the document portal in flatpak, but listen to @hadess. He knows stuff :-)

**Comment by kensully** [comment]
@mariospr @hadess I rebuilt all the packages with different options provided by `./configure --help` pointint at system paths like 

        ./configure --with-dbus-service-dir=/usr/share/dbus-1/services --with-dbus-config-dir=/etc/dbus-1/system.d --with-systemduserunitdir=/usr/lib/systemd/user --with-systemdsystemunitdir=/usr/lib/systemd/system

Above configuration installs `dbus` and `systemd` stuff onn default system path but it installs applications into `/usr/local`, but it didn't work. So I uninstalled them all then reconfigured them all flatpak including its dependencies with `./configure --prefix=/usr`, it must have definitely worked, but it didn't. 

It occured to me that it worked well in `/usr/local` path well, so it can not be a path issue since applications from `flathub` has no issue regarding the file choose dialog. 

Finally I checked out the messages when I click `open file`  on the application with `dbus-monitor

    method call time=1520432678.844619 sender=:1.203 -> destination=org.freedesktop.portal.Desktop serial=19 path=/org/freedesktop/portal/desktop; interface=org.freedesktop.portal.FileChooser; member=OpenFile
       string "x11:117440517"
       string "Open Vocabulary Document"
       array [
          dict entry(
             string "filters"
             variant             array [
                   struct {
                      string "KDE Vocabulary Document "
                      array [
                         struct {
                            uint32 0
                            string "*.kvtml"
                         }
                      ]
                   }
                   struct {
                      string "KWordQuiz Document "
                      array [
                         struct {
                            uint32 0
                            string "*.wql"
                         }
                      ]
                   }
                   struct {
                      string "Pauker Lesson "
                      array [
                         struct {
                            uint32 0
                            string "*.xml.qz"
                         }
                         struct {
                            uint32 0
                            string "*.pau.gz"
                         }
                      ]
                   }
                   struct {
                      string "Vokabeltrainer "
                      array [
                         struct {
                            uint32 0
                            string "*.voc"
                         }
                      ]
                   }
                   struct {
                      string "XML Dictionary Exchange Format "
                      array [
                         struct {
                            uint32 0
                            string "*.xdxf"
                         }
                      ]
                   }
                   struct {
                      string "Comma Separated Values (CSV) "
                      array [
                         struct {
                            uint32 0
                            string "*.csv"
                         }
                      ]
                   }
                ]
          )
          dict entry(
             string "modal"
             variant             boolean true
          )
          dict entry(
             string "multiple"
             variant             boolean true
          )
       ]
    error time=1520432678.859806 sender=org.freedesktop.DBus -> destination=:1.203 error_name=org.freedesktop.DBus.Error.Spawn.ChildExited reply_serial=19
       string "Process org.freedesktop.portal.Desktop exited with status 127"

It appears to be not using `xdg-desktop-portal-kde` and `xdg-desktop-portal` not compatible with `kde` applications in flatpak.

**Comment by kensully** [comment]
https://github.com/flatpak/flatpak/issues/1276
